### DISCUSSION TITLE
the use of displaced mesh

### DISCUSSION THREAD

**Discussion by xiaojbing** [idaholab/moose#20583] [Q&A Meshing] [answered]
I want to implement the displaced mesh configuration by using the "displacement='disp_x disp_y'" and "_use_displaced_mesh=true" . The mesh can now be seen moving over time on the console, but the solve show the same results as if the displacement field had not been applied. I suspect that the coordinate values do not change at each step, resulting in results that are not related to the moving mesh. Is there any way that the coordinate values can be changed at each step to affect the calculation of the equations.

```
#[GlobalParams]
  #order = FIRST
  #family = LAGRANGE
  #disp_x = disp_x
  #disp_y = disp_y
  #displacements = 'disp_x disp_y'
#[]

[GlobalParams]
  displacements = 'disp_x disp_y'
  #use_displaced_mesh = true
[]

[Mesh]
  file = grids/test03.exo
  use_displaced_mesh = true
  displacements = 'disp_x disp_y'

  block_id = 3
  block_name = 'fluid'

  boundary_id = '1 2'
  boundary_name = 'wall far'
  uniform_refine = 1
[]

[Problem]
  type = EulerProblem
  mach = 0.6
  attack = 0
  jacobian_delay = 1
  flux_type = HLL

  [./Kernels]
    use_displaced_mesh = true
    displacements = 'disp_x disp_y'
    type = CFDCellKernel
    shock_indicator = indicator
    block = '3'
  [../]

  [./DGKernels]
    type = CFDFaceKernel
    block = '3'
  [../]
[]

[AuxVariables]
  [./disp_x]
    order = FIRST
    family = LAGRANGE
    block = '3'
  [../]

  [./disp_y]
    order = FIRST
    family = LAGRANGE
    block = '3'
  [../]
[]

[AuxKernels]
  [./disp_x_kernel]
    use_displaced_mesh = true
    type = FunctionAux
    variable = disp_x
    function = 'sqrt(x^2+y^2)*cos(atan(y/x)-4.86/360*2*pi-2.44*sin(0.1620*t)/360*2*pi) - x'
    block = '3'
  [../]

  [./disp_y_kernel]
    use_displaced_mesh = true
    type = FunctionAux
    variable = disp_y
    function = 'sqrt(x^2+y^2)*sin(atan(y/x)-4.86/360*2*pi-2.44*sin(0.1620*t)/360*2*pi) - y'
    block = '3'
  [../]
[]

[Variables]
  order = CONSTANT
  family = MONOMIAL
  variables = 'density momx momy momz rhoe'
[]

[ICs]
  type = CFDPassFlowIC
[]

[BoundaryCondition]
  [./euler_far_field]
    #use_displaced_mesh = true
    type = FarFieldPressure
    boundary = '2'
  [../]

  [./euler_wall]
    #use_displaced_mesh = true
    type = RotarySlipWall
    boundary = '1'
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Adaptivity]
  #marker = marker
  #max_h_level = 2
  [./Indicators]
    [./indicator]
      type =  DensityJumpIndicator
      variable = density
      scale = 0
    [../]
  [../]
  [./Markers]
    [./marker]
      type = ErrorMaxFractionMarker
      indicator = indicator
      coarsen = 0.7
      refine = 0.5
    [../]
  [../]
[]


[Executioner]
  no_fe_reinit = true
  type = Transient
  solve_type = NEWTON
  num_steps = 1000
  l_tol = 5e-02
  l_max_its = 20

  nl_max_its = 20
  nl_rel_tol = 0.999999

  [./TimeStepper]
    type = RatioTimeStepper
    dt = 1E-02
    ratio = 2
    step = 5
    max_dt = 0.5
  [../]
[]

[CFDForce]
  [cfdforce]
    #use_displaced_mesh = true
    type = CFDForceUserObject
    boundary = 'wall'
    ref_point = '0 0 0'
    ref_length = 1
  []
[]

[Postprocessors]
  [./residuals]
    type = VariableResidual
    variable = density
  [../]
  [cd]
    type = CFDForcePostprocessor3
    solution = cfdforce
  []
  [cl]
    type = CFDForcePostprocessor2
    solution = cfdforce
  []
[]

[Outputs]
  csv = true
  checkpoint = true
  #show = 'density momx momy momz rhoe disp_x disp_y'

  [./console]
    type = Console
    perf_log = true
    execute_on = 'linear nonlinear'
  [../]

  [./exodus]
    type = Exodus
    execute_on = 'initial timestep_end'
    interval = 2
  [../]

  #[./tecplot]
    #type = Tecplot
    #execute_on = 'initial timestep_end'
    #interval = 10
  #[../]
[]
```

![116a688080ac6483093f95fd053338b](https://user-images.githubusercontent.com/87591264/158820441-fece4e4b-8ca1-4a97-886e-3d19a4eb6ee6.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

I assume that by the `cloud` you mean the exodus output?
Have you tried to output the displaced mesh? For this you will need an exodus block in Outputs with the `use_displaced` parameter set to true. 
```
[Outputs]
  [exo]
    type = Exodus
    use_displaced = true
  []
[]
```

Guillaume

**Reply by xiaojbing** [reply]
Thank you so much. But I have another question. When I use the  "use_displaced_mesh=true" in the [DGKernel] block, I get an error when I run it, but I found an example inputfile shows that this parameter can be used in the block, so what is the reason for this error?

![cb856a074411a6cb994d17e3a5a40ed](https://user-images.githubusercontent.com/87591264/159004368-e0ffd7a4-09ef-48a7-a047-052992ea7c6b.png)

**Reply by GiudGiud** [reply]
Hello

So you hit a seg fault. You have narrowed the issue down to `CFDFaceKernel` ?
Which application is that from? I dont see it in the MOOSE documentation.

Guillaume

**Reply by xiaojbing** [reply]
In my input file, I found that the` displacements='disp_x disp_y disp_z'` in the `[mesh]` block and the `use_displaced_ mesh=true` in the
`[DGkernels]` block would conflict, and if both sytax were added at the same time, an error would appear on the console, whereas if the` true` was changed to `false`, it would work fine.
![image](https://user-images.githubusercontent.com/87591264/160401646-e31706d0-cc4e-4f50-a3de-8f525b250a2e.png)
I found that the reason may be that` use displaced mesh` is false by default, which causes it not to run after I set it to true in the input
file, and I wonder if there is a way to change this mandatory-like parameter.
![image](https://user-images.githubusercontent.com/87591264/160401914-cc41ab78-1bd7-40b0-8fb8-f3d6c7102342.png)
![image](https://user-images.githubusercontent.com/87591264/160402729-2f7a8309-7ad3-4b9d-8312-be799552d048.png)

**Reply by GiudGiud** [reply]
`ADDGDiffusion` works fine with displaced meshes. There is an example here:
https://github.com/idaholab/moose/blob/next/test/tests/dgkernels/dg_displacement/dg_displacement.i

I recommend you use our debugging workflow here
https://mooseframework.inl.gov/application_development/debugging.html
to isolate the problem down to a kernel. I suspect the CFDFaceKernel.

**Reply by xiaojbing** [reply]
thank you...  But I find that ADDGDiffusion inherits from `DGKernel`, and the `CFDFaceKernel` what I am using also inherits from DGKernel. In the dg_displacement.i, the `ADDGDiffusion` have not used the paramete "boundary = " , and in this case it could use the paramete `use_displaced_mesh = true` . But in my input file. I have to use the paramete `boundary =`  to run the `use_displaced_mesh = true` , but this also leads to unreasonable results.
![image](https://user-images.githubusercontent.com/87591264/162424532-b594bd9a-7fb1-435b-8b95-9377d68fdc6b.png)
![image](https://user-images.githubusercontent.com/87591264/162424572-b1a15098-7a5c-45a6-936c-c9fa7d6c3682.png)

**Reply by GiudGiud** [reply]
This is a problem with your implementation of this DGKernel.
Please see these instructions for debugging. 
https://mooseframework.inl.gov/application_development/debugging.html
once you have obtained a backtrace in debug mode for this error we may be able to help you

Do not add `boundary =` if the kernel is to be executed in the volume

**Reply by xiaojbing** [reply]
Thank you for your patience. I can run this simulation now. But I still hava some qustions to ask. 
I create an application named parakeet to slove CFD problems, and I add my own physics in it. Now I want to use displaced mesh in input file. 

In parakeet, The `CFDCellKernel` inherit Kernel.C. The `CFDFaceKernel` inherit DGKernel.C. The `CFDBC` inherit IntegratedBC.C. 

I set the displacement field to 0 and then used `use_displaced_mesh=true`  only in [Kernel], [DGKernel] and [BCs] respectively. The expected results should be same as if I don't add displacement. But  the results are reasonable when I only use it in [kernel], the other two situation are not as expected. 

I had tried to annotation code to find the reason. And I found that when the code in the red box was removed in the CFDFaceKernel.CPP and CFDBC.CPP, the results will be reasonable. But the removed code  is only related to the time-marching, which should only affect convergence,  not calculation results. 

I have also tried adjusting the convergence exponent to reduce the residuals to a very low level and the results are still unreasonable. So the reason for the unreasonable results is not convergence. Am I overlooking something, is there anything else I need to pay special attention to when using displaced mesh?

CFDFaceKernel.cpp & CFDBC.cpp:
![image](https://user-images.githubusercontent.com/87591264/163946363-850597e5-890e-4ef3-a9ca-21811e7d58c0.png)

input file:
```
[Mesh]
  displacements = 'disp_x disp_y'
  file = grids/test03.exo
[]

[Problem]
  type = EulerProblem
  mach = 0.15

  [./Kernels]
    disp_x = disp_x
    disp_y = disp_y
    #use_displaced_mesh = true
    type = CFDCellKernel
  [../]

  [./DGKernels]
    disp_x = disp_x
    disp_y = disp_y
    #use_displaced_mesh = true
    type = CFDFaceKernel
  [../]
[]

[AuxVariables]
  [./disp_x]
    order = FIRST
    family = LAGRANGE
  [../]
  [./disp_y]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[AuxKernels]
  [./disp_x_kernel]
    type = FunctionAux
    variable = disp_x
    function = '0'
  [../]
  [./disp_y_kernel]
    type = FunctionAux
    variable = disp_y
    function = '0'
  [../]
[]

[Variables]
  order = CONSTANT
  family = MONOMIAL
  variables = 'density momx momy momz rhoe'
[]

[ICs]
  type = CFDPassFlowIC
[]

[BoundaryCondition]
  [./euler_far_field]
    use_displaced_mesh = true
    type = FarFieldPressure
    boundary = far
  [../]

  [./euler_wall]
    use_displaced_mesh = true
    type = SlipWall
    boundary = wall
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
    solve_type = 'NEWTON'
  [../]
[]

[Executioner]
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  type = Transient
  solve_type = PJFNK
  scheme = bdf2

  line_search = 'none'
  nl_abs_tol = 1e-20
  start_time = 0.0
  timestep_tolerance = 1e-6

  num_steps = 1000
  l_tol = 1e-10
  l_max_its = 20

  nl_max_its = 20
  nl_rel_tol = 1e-10

  dt = 0.01
  end_time = 5.0
[]

[Outputs]
  csv = true
  checkpoint = true
  [./console]
    type = Console
    perf_log = true
    execute_on = 'linear nonlinear'
  [../]
  [./exodus]
    type = Exodus
    execute_on = 'initial timestep_end'
  [../]
[]

```

**Reply by GiudGiud** [reply]
Honestly not sure. 
I would look at tensor_mechanics and examine how they use the displaced variables in their FE solve.
If you can find other examples of DGKernels adapted for the displaced mesh that's also worth taking into consideration.

A 0 displacement should indeed not affect the solve

**Reply by xiaojbing** [reply]
Thank you very much, I will try.

**Reply by GiudGiud** [reply]
@xiaojbing any update on this?
Did you find the problem?
